src/main/java/com/shapesecurity/functional/data/ImmutableList.java

```diff
@@ -75,31 +76,59 @@
     }
 
     /**
-     * Creating List from.
+     * Creating ImmutableList from an ArrayList.
      *
-     * @param arrayList The {@link java.util.ArrayList} to construct the {@link ImmutableList}
+     * @param list      The {@link java.util.ArrayList} to construct the {@link ImmutableList}
      *                  from.
      * @param <A>       The type of the elements of the list.
      * @return a new {@link ImmutableList} that is comprised of all the elements in the {@link
      * java.util.ArrayList}.
      */
     @Nonnull
-    public static <A> ImmutableList<A> from(@Nonnull List<A> arrayList) {
-        // Manual expansion of tail recursion.
+    public static <A> ImmutableList<A> from(@Nonnull ArrayList<A> list) {
         ImmutableList<A> l = empty();
-        int size = arrayList.size();
+        int size = list.size();
         for (int i = size - 1; i >= 0; i--) {
-            l = cons(arrayList.get(i), l);
+            l = cons(list.get(i), l);
         }
         return l;
-    /* TODO: use Collection here if it doesn't incur performance penalties
-    ImmutableList<A> list = empty();
-    int size = collection.size();
-    for (A element : collection) {
-      list = cons(element, list);
     }
-    return list;
+
+    /**
+     * Creating ImmutableList from a Deque. Generally used for {@link java.util.LinkedList}, but applicable to reverse iterable types in general.
+     *
+     * @param list      The {@link java.util.Deque} to construct the {@link ImmutableList}
+     *                  from.
+     * @param <A>       The type of the elements of the list.
+     * @return a new {@link ImmutableList} that is comprised of all the elements in the {@link
+     * java.util.Deque}.
+     */
+    @Nonnull
+    public static <A> ImmutableList<A> from(@Nonnull Deque<A> list) {
+        ImmutableList<A> l = empty();
+        for (Iterator<A> iterator = list.descendingIterator(); iterator.hasNext();) {
+            A item = iterator.next();
+            l = cons(item, l);
+        }
+        return l;
+    }
+
+    /**
+     * Creating ImmutableList from an Iterable.
+     *
+     * @param list      The {@link java.lang.Iterable} to construct the {@link ImmutableList}
+     *                  from.
+     * @param <A>       The type of the elements of the list.
+     * @return a new {@link ImmutableList} that is comprised of all the elements in the {@link
+     * java.util.ArrayList}.
      */
+    @Nonnull
+    public static <A> ImmutableList<A> from(@Nonnull Iterable<A> list) {
+        ImmutableList<A> l = empty();
+        for (A item : list) {
+            l = cons(item, l);
+        }
+        return l.reverse();
```

Comment:
I think it's likely to be faster to build an ArrayList and then then call the `from(ArrayList)` overload, no?